### PR TITLE
Rework MaeMolSupplier, fix #2617

### LINKED ISSUE #2617: Chem.rdmolfiles.MaeMolSupplier Never Stops Reading
**Description:**

- RDKit Version: 2019.03.01
- Platform: Linux

`Chem.rdmolfiles.MaeMolSupplier` reads more molecules than is actually in the file.
If you try to read until file exhaustion it never stops reading.

```
################### Schrodinger Suite Reads the File #######################
from schrodinger import structure
from rdkit import Chem
import datetime
 
reader = structure.StructureReader('pm7_1212_0_out.mae')
sts = [x for x in reader]
print("total molecules = %s" % len(sts))
total_mols = len(sts)
# total molecules = 5225
 
 ################### rdkit will just keep reading forever #######################

suppl = Chem.rdmolfiles.MaeMolSupplier('pm7_1212_0_out.mae')
start_time = datetime.datetime.now()
mols = []
for i, m in enumerate(suppl):
  mols.append(m)
  if i > total_mols:
    print("MaeMolSupplier Keeps Reading Molecules after all are gone")
    break
# MaeMolSupplier Keeps Reading Molecules after all are gone

```
[example mae file](https://karlleswing.com/misc/pm7_1212_0_out.mae)

### PR DESCRIPTION
**Pull request by ricrogz**
This started as a fix for #2617, but I added some more changes. That issue is caused by #2579, but after finding the problem, the `MaeMolSupplier` does not stop iteration despite it does not get any more structures, and continues returning "None" forever.

Also, this depends on schrodinger/maeparser#46 and schrodinger/maeparser#49, and should not be merged until these changes are accessible to RDKit.

Changes in the PR:

- Check the state of the stream after the creation of the Supplier: on construction, the `Reader` populates the internal buffer, so that, if there aren't any problems, the stream may be left either in a `good` state, if there was more data available than fits in the internal buffer, or with the `failed` and `eof` bits activated if there was less data than the buffer tried to read (both states are ok).

- Do not read the next block / structure before it is requested. Before this change, the MaeMolSupplier attempted to read the first structure still inside the constructor, right after constructing the Reader object; the second one when the first was requested, etc. With this mechanism, if a problem was encountered while reading the second structure, the first one would not have been returned.

- Catching of exceptions thrown during and after reading of the `mae::Block`, and rethrowing them as `FileParseException` so that they can be handled by RDKit (e.g., by preventing further reads, like should happen in #2617).

- Changed the way how `atEnd()` is handled. This depends on schrodinger/maeparser#49.

- Provide translation of `FileParseException`s into Python `RuntimeErrors`. This is the part that would actually prevent an infinite reading loop as in #2617, and also the one I am most interested in, especially because of the comment at:
https://github.com/rdkit/rdkit/blob/01fbec33ff434432d840828b7fb02cc22680fd42/Code/GraphMol/Wrap/MolSupplier.h#L90-L94
I tried to iterate over a `MaeMolSupplier`, a `SDMolSupplier` and a `ForwardSDMolSupplier` after introducing this patch, and did not observe any problems like unexpected `RuntimeError`s, e.g. at the end of the iteration).

### CODE REVIEW

**Comment by ricrogz** [comment]
Updated to parse chirality, pseudochirality and stereo bond properties instead of just copying them to the mol, and updated to current master again.

This adds a new dependence on another maeparser PR, schrodinger/maeparser/pull/50, which adds constants for the prefixes used in stereo properties, and some comment on how these properties are build.

**Comment by ricrogz** [comment]
I removed the dependence on the `eof()` method in maeparser, as we had certain reservations on it.

**Comment by ricrogz** [comment]
Merged in current master (after maeparser and coordgen update). This should build fine now, as  schrodinger/maeparser/pull/50 has also been merged

**Comment by greglandrum** [comment]
@ricrogz : those build failures on windows look real. Can you look into them?

**Comment by ricrogz** [comment]
> @ricrogz : those build failures on windows look real. Can you look into them?

Yeah, sorry, I noticed, but something came up before I could look into it, but will do soon.

**Comment by ricrogz** [comment]
Ok, now it should finally build and pass tests on all platforms :)

**Comment by ricrogz** [comment]
I realized this had become too messy, and have reset the original branch.

File: Code/GraphMol/FileParsers/MaeMolSupplier.cpp [hunk 1]
```diff
@@ -10,243 +10,424 @@
...
+const std::string PDB_CHAIN_NAME = "s_m_chain_name";
+const std::string PDB_INSERTION_CODE = "s_m_insertion_code";
+const std::string PDB_RESIDUE_NUMBER = "i_m_residue_number";
+const std::string PDB_OCCUPANCY = "r_m_pdb_occupancy";
+const std::string PDB_TFACTOR = "r_m_pdb_tfactor";
 
-  d_reader.reset(new Reader(dp_sInStream));
-  d_next_struct = d_reader->next("f_m_ct");
-}
+class PDBInfo {
```
Comment (d-b-w) [inline Code/GraphMol/FileParsers/MaeMolSupplier.cpp]: I'm a bit confused by this diff - this part seems like it's already merged in. I'm just trying to separate out the two fixes in my head.

File: Code/GraphMol/FileParsers/MaeMolSupplier.cpp [hunk 2]
```diff
@@ -10,243 +10,424 @@
...
+  boost::tokenizer<boost::char_separator<char>> tokenizer{stereo_prop, sep};
+
+  auto tItr = tokenizer.begin();
+
+  const int chiral_idx = FileParserUtils::toInt(*tItr) - 1;
+  Atom *chiral_atom = mol.getAtomWithIdx(chiral_idx);
+  CHECK_INVARIANT(chiral_atom != nullptr, "bad prop value");
+
+  unsigned nSwaps = 2;
+  switch (*((++tItr)->rbegin())) {
```
Comment (d-b-w) [inline Code/GraphMol/FileParsers/MaeMolSupplier.cpp]: Does the token have a `.back()` like `vector` and `string`?

File: Code/GraphMol/FileParsers/MaeMolSupplier.cpp [hunk 3]
```diff
@@ -10,243 +10,424 @@
...
+  INT_LIST bond_indexes;
+  for (++tItr; tItr != tokenizer.end(); ++tItr) {
+    const int nbr_idx = FileParserUtils::toInt(*tItr) - 1;
+    const Bond *bnd = mol.getBondBetweenAtoms(chiral_idx, nbr_idx);
+    bond_indexes.push_back(bnd->getIdx());
+  }
+  CHECK_INVARIANT(bond_indexes.size() == chiral_atom->getDegree(),
+                  "bad prop value");
+
+  nSwaps += chiral_atom->getPerturbationOrder(bond_indexes);
```
Comment (d-b-w) [inline Code/GraphMol/FileParsers/MaeMolSupplier.cpp]: What about hydrogens? If this is a chirality, any hydrogen would be at the end. But if this were a molecule with explicit hydrogens and atom-numbering-only absolute configuration, the hydrogen wouldn't necessarily be last in the mae property.

File: Code/GraphMol/FileParsers/MaeMolSupplier.cpp [hunk 4]
```diff
@@ -10,243 +10,424 @@
...
 
-    if(pdb_info.m_resnum && pdb_info.m_resnum->isDefined(atom_num))
-    {
-        rd_info->setResidueNumber(pdb_info.m_resnum->at(atom_num));
+    atom.setProp(prop.first, prop.second->at(i));
+  }
+
+  for (const auto &prop : atom_block.getProperties<double>()) {
+    if (prop.first == mae::ATOM_X_COORD || prop.first == mae::ATOM_Y_COORD ||
+        // Coordinates are used in defining a conformation, and should not be
```
Comment (d-b-w) [inline Code/GraphMol/FileParsers/MaeMolSupplier.cpp]: weird place for the comment, please move it either into the block or out of it.

File: Code/GraphMol/FileParsers/MaeMolSupplier.cpp [hunk 5]
```diff
@@ -10,243 +10,424 @@
...
+}
 
-    if(pdb_info.m_insertion_code &&
-            pdb_info.m_insertion_code->isDefined(atom_num))
-    {
-        rd_info->setInsertionCode(pdb_info.m_insertion_code->at(atom_num));
+//! Set atom properties. Some of these have already been parsed to construct the
+//! Atom object, and should be skipped. Also, atom properties may be undefined
+//! for an atom, and should also be skipped for that atom.
+void set_atom_properties(Atom &atom, const mae::IndexedBlock &atom_block,
```
Comment (d-b-w) [inline Code/GraphMol/FileParsers/MaeMolSupplier.cpp]: probably more efficient to set a single property on all atoms and then move on to the next property. We can leave that for later, though.

File: Code/GraphMol/FileParsers/MaeMolSupplier.cpp [hunk 6]
```diff
@@ -10,243 +10,424 @@
...
     }
 
-    if(pdb_info.m_occupancy && pdb_info.m_occupancy->isDefined(atom_num))
-    {
-        rd_info->setOccupancy(pdb_info.m_occupancy->at(atom_num));
+    atom.setProp(prop.first, prop.second->at(i));
+  }
+  for (const auto &prop : atom_block.getProperties<int>()) {
+    if (prop.first == mae::ATOM_ATOMIC_NUM) {
+      // Atomic number was already used in the creation of the atom
```
Comment (d-b-w) [inline Code/GraphMol/FileParsers/MaeMolSupplier.cpp]: Should we delete the property when we consume it?

File: Code/GraphMol/FileParsers/MaeMolSupplier.cpp [hunk 7]
```diff
@@ -10,243 +10,424 @@
...
-    const auto zs = atom_data->getRealProperty("r_m_z_coord");
-    const auto atom_count = atomic_numbers->size();
-    shared_ptr<IndexedIntProperty> atomic_charges;
-    try {
-      atomic_charges = atom_data->getIntProperty("i_m_formal_charge");
-    } catch (std::out_of_range &) {
-    }
+void addAtoms(const mae::IndexedBlock &atom_block, RWMol &mol) {
+  // All atoms are guaranteed to have these three field names:
+  const auto atomic_numbers = atom_block.getIntProperty(mae::ATOM_ATOMIC_NUM);
```
Comment (d-b-w) [inline Code/GraphMol/FileParsers/MaeMolSupplier.cpp]: not quite true. All atoms have either atomic number or atom type (or both).

File: Code/GraphMol/FileParsers/MaeMolSupplier.cpp [hunk 8]
```diff
@@ -10,243 +10,424 @@
...
+  const auto to_atoms = bond_block.getIntProperty(mae::BOND_ATOM_2);
+  const auto orders = bond_block.getIntProperty(mae::BOND_ORDER);
+  const auto size = from_atoms->size();
+
+  for (size_t i = 0; i < size; ++i) {
+    // Maestro atoms are 1 indexed!
+    const auto from_atom = from_atoms->at(i) - 1;
+    const auto to_atom = to_atoms->at(i) - 1;
+    const auto order = bolookup.find(orders->at(i))->second;
+    if (from_atom > to_atom) continue;  // Maestro files double-list bonds
```
Comment (d-b-w) [inline Code/GraphMol/FileParsers/MaeMolSupplier.cpp]: _Older_ Maestro files double-list bonds.

File: Code/GraphMol/FileParsers/MaeMolSupplier.cpp [hunk 9]
```diff
@@ -10,243 +10,424 @@
...
+
+    auto bond = new Bond(order);
+    bond->setOwningMol(mol);
+    bond->setBeginAtomIdx(from_atom);
+    bond->setEndAtomIdx(to_atom);
+    mol.addBond(bond, true);
```
Comment (d-b-w) [inline Code/GraphMol/FileParsers/MaeMolSupplier.cpp]: I've only used:

    mol.addBond(from_atom, to_atom, order);

Is there a reason to prefer one idiom over the other?

File: Code/GraphMol/FileParsers/MaeMolSupplier.cpp (hunk 1 above)
Comment (ricrogz) [reply Code/GraphMol/FileParsers/MaeMolSupplier.cpp]: Ah, I think I know what the problem is: I created the PR, and after that, the `PDBInfo` was added, and I had to update the PR to include it too. Is it possible that it is that?

File: Code/GraphMol/FileParsers/MaeMolSupplier.cpp (hunk 2 above)
Comment (ricrogz) [reply Code/GraphMol/FileParsers/MaeMolSupplier.cpp]: It is actually a string, so it does. I'll make this a bit clearer.

File: Code/GraphMol/FileParsers/MaeMolSupplier.cpp (hunk 4 above)
Comment (ricrogz) [reply Code/GraphMol/FileParsers/MaeMolSupplier.cpp]: Yes, that one is quite weird. Not sure how it got there, though.

File: Code/GraphMol/FileParsers/MaeMolSupplier.cpp (hunk 5 above)
Comment (ricrogz) [reply Code/GraphMol/FileParsers/MaeMolSupplier.cpp]: I actually tried both options with just a few properties, and I didn't see too much of a difference. But, as you say, we might come back to this later on.

File: Code/GraphMol/FileParsers/MaeMolSupplier.cpp (hunk 6 above)
Comment (ricrogz) [reply Code/GraphMol/FileParsers/MaeMolSupplier.cpp]: I don't really have a preference about that. I don't think it is going to be a big performance improvement, so maybe we can just leave it be, and do it together with the other change about setting one property for all atoms at a time.

File: Code/GraphMol/FileParsers/MaeMolSupplier.cpp (hunk 7 above)
Comment (ricrogz) [reply Code/GraphMol/FileParsers/MaeMolSupplier.cpp]: Yeah, but we don't have a way of translating to/from atom types without Schrodinger libraries, so we can't handle atom types, which means that need atomic numbers.

File: Code/GraphMol/FileParsers/MaeMolSupplier.cpp (hunk 8 above)
Comment (ricrogz) [reply Code/GraphMol/FileParsers/MaeMolSupplier.cpp]: They still do if the bond has properties. I guess Maestro files double-list *some* bonds.

File: Code/GraphMol/FileParsers/MaeMolSupplier.cpp (hunk 9 above)
Comment (ricrogz) [reply Code/GraphMol/FileParsers/MaeMolSupplier.cpp]: `mol.addBond(from_atom, to_atom, order);` can't take indexes for `from_atom` and `to_atom`, they have to be `Atom*`.

File: Code/GraphMol/FileParsers/MaeMolSupplier.cpp (hunk 3 above)
Comment (ricrogz) [reply Code/GraphMol/FileParsers/MaeMolSupplier.cpp]: At the point of doing this, all atoms are still explicit: in the `next()` method, first, atoms are parsed, then bonds, then mol properties, and then, Hydrogen nodes are removed, if we opted to do so.

**Pull request merged**